etc/docker-scripts/bootstrapper.py

```diff
@@ -118,6 +122,150 @@ def process_outputs(self) -> None:
         duration = time.time() - t0
         OpUtil.log_operation_info('outputs processed', duration)
 
+    def process_metrics(self) -> None:
+        """Expose metrics and    metadata in the KFP UI
+
+        When invoked this method produces a
+        https://www.kubeflow.org/docs/pipelines/sdk/output-viewer
+        compatible metadata file, which esposes the following information
+        for each pipeline node in the KFP UI:
+         - the name of the input file (e.g. notebook.ipynb)
+         - the name of the input archive (e.g. notebook...tar.gz)
+         - the location of the input archive on COS
+
+        If the notebook | script produced a metrics file it is also
+        exposed in the KFP UI:
+        https://www.kubeflow.org/docs/pipelines/sdk/pipelines-metrics/
+
+        There should be no need to override this method in subclasses
+        because the code is not file type specific.
+        """
+
+        OpUtil.log_operation_info('processing metrics and metadata')
+        t0 = time.time()
+
+        # Location where the KFP specific output files will be stored
+        output_path = Path('/tmp')
```

Comment:
Should this be something like [`Path(tempfile.gettempdir())`](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir) in order to address platform differences?

Reply:
In theory it should, but it's a bit of a chicken and egg problem that I was dealing with. The exact path needs to be known before the bootstrapper is invoked. A bit of background information: 
- In https://github.com/elyra-ai/elyra/pull/1054/files we declare two output files, which are used to make small amounts of data available to KFP
- if one declares `mlpipeline-metrics` or `mlpipeline-ui-metadata` as `output_files` for a KFP component, KFP will look for those files in the running container.
- if a declared file is found it's content is rendered in the KFP UI

If we were to use a runtime-based approach to determine the temp directory location that might yield different results (e.g. `/temp`), the declaration would be no longer valid and KFP unable to locate the files. I couldn't think of any well defined directory location that would exist (and be writable) in every container image and is not platform dependent. Therefore I went with the static directory choice, which isn't great (and would certainly lead to failure if one were to run a notebook | script on a windows-based container image).

I haven't tried out what happens if one does not specify a path in the declaration at all (aka `mlpipe...json`) since I couldn't determine what `cwd` would  resolve to in the running container, and whether the result would apply to all containers.

Reply:
What I can do if there's no solution to the problem is to add defensive code that checks whether the specified static location exists and is writable. If it is not then the method would essentially act like a no-op and the metadata would not be exposed in the KFP UI, avoiding failure.

Reply:
> if one declares mlpipeline-metrics or mlpipeline-ui-metadata as output_files for a KFP component

Isn't that what the linked elyra PR is doing unconditionally?  Where's the 'if' here?  I'm confused.   And how will the code in the elyra PR know what kind of filesystem (or its variances) the underly image will have?  Seems like the best we can do (and should do) is convey the filename, not the location.  You could then look for these particular named files in the specified outputs to get their locations as they may not be in `'.'`. 

I'm sorry that I'm not more up on how this works as I suspect I have a disconnect here.

> What I can do if there's no solution to the problem is to add defensive code that checks whether the specified static location exists and is writable. If it is not then the method would essentially act like a no-op

Based on @akchinSTC's experiences with ODH, I think this defensive code is going to be necessary as I believe even `/tmp` is not writable and a special mounted FS is required.  (Alan, could you please confirm?)

(Which it just occurs to me that all this code is in a container and I don't think there will be containers running Windows in this context, so my concern about platform specifics in my other comment is not as worrisome, although different flavors of *nix do handle temporary directories differently, etc.)

Reply:
> Isn't that what the linked elyra PR is doing unconditionally?

Yes, we have to because of a similar chicken and egg problem - we don't know ahead of time whether a notebook | Python script will produce a metadata file or not.

> And how will the code in the elyra PR know what kind of filesystem (or its variances) the underly image will have?

We do not know.

> convey the filename, not the location. You could then look for these particular named files in the specified outputs to get their locations as they may not be in '.'

The problem is that we don't know how and where KFP will look for the file(s) if we don't specify an absolute path when we declare the `output_files`. Therefore we don't know where to store them at runtime.

From the KFP doc: "If the component writes such a file to its container filesystem, the Kubeflow Pipelines system extracts the file, and the Kubeflow Pipelines UI uses the file to generate the specified viewer(s)."

> this defensive code is going to be necessary

I just added it.